### PR TITLE
Disable crawlers through config

### LINKED ISSUE #70: Allow crawlers to be disabled
**Is your feature request related to a problem? Please describe.**

There is no simple way to prevent some crawlers for being executed. It would be nice to have some way to disable misbehaving crawlers o just disable a specific crawler during development.

**Describe the solution you'd like**

A way to disable specific crawlers through config.

**Describe alternatives you've considered**

The current alternative is to comment crawlers code.

**Additional context**

### PR DESCRIPTION
**Pull request by Bigomby**
**Description:**

Improve configuration to allow disabling some crawlers.

Also fix some calls to `setTimeout` where the argument is not a function.

---

For contributor:

- [x] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

---

Closes #70

### CODE REVIEW

File: backend.config.js [hunk 1]
```diff
@@ -1,20 +1,44 @@
...
     database: 'polkastats',
     password: 'polkastats',
     port: 5432,
   },
-  BLOCK_HARVESTER_POLLING_TIME: 1 * 60 * 1000, // Run every 10 min
-  ACTIVE_ACCOUNTS_POLLING_TIME: 1 * 60 * 1000, // Run every 60 min
-  PHRAGMEN_POLLING_TIME: 5 * 60 * 1000, // Run every 5 min
+
+  crawlerRunner: {
+    wsProviderUrl: process.env.WS_PROVIDER_URL || DEFAULT_WS_PROVIDER_URL,
```
Comment (mariopino) [inline backend.config.js]: Why we need to include `wsProviderUrl` in all crawlers?

I would use:

```
// Also wss://kusama-rpc.polkadot.io
const DEFAULT_WS_PROVIDER_URL = 'wss://substrate-node:9944';

module.exports = {
  postgresConnParams: {
    user: 'polkastats',
    host: 'postgres',
    database: 'polkastats',
    password: 'polkastats',
    port: 5432,
  },
  wsProviderUrl: process.env.WS_PROVIDER_URL || DEFAULT_WS_PROVIDER_URL,
  crawlers: {
    blockListener: {
      enabled: true,
      startingDelay: false
    },
    blockHarvester: {
      enabled: true,
      startingDelay: true,
      pollingTime: 1 * 60 * 1000,
    },
    staking: {
      enabled: true,
      startingDelay: false
    },
    rewards: {
      enabled: true,
      startingDelay: false
    },
    accounts: {
      enabled: true,
      startingDelay: true,
      pollingTime: 1 * 60 * 1000,
    },
    phragmen: {
      enabled: true,
      startingDelay: true,
      pollingTime: 5 * 60 * 1000,
      phragmenOutputDir: '/tmp/phragmen',
      offlinePhragmenPath: '/usr/app/polkastats-backend-v3/offline-phragmen',
    }
  }
};
```

File: lib/BackendV3.js [hunk 2]
```diff
@@ -14,50 +14,48 @@ class BackendV3 {
...
+      activeAccounts(api, pool, this.config.accounts);
+    }
+
+    if (this.config.staking) {
+      staking(api, pool, this.config.staking);
+    }
+
+    if (this.config.phragmen) {
+      phragmen(api, pool, this.config.phragmen);
+    }
```
Comment (Bitcoinera) [inline lib/BackendV3.js]: This one looks definitely much cleaner and readable than the old version.

Just wondering if there is a bit less repetitive way to put it, but I think this is the way to go

File: lib/crawlers/activeAccounts.js [hunk 3]
```diff
@@ -55,6 +54,10 @@ module.exports = {
         } 
       }
     }
-    setTimeout(module.exports.activeAccounts(api, pool, config), POLLING_TIME);
+
+    setTimeout(
+      () => module.exports.activeAccounts(api, pool, config),
+      config.pollingTime,
+    );
```
Comment (Bitcoinera) [inline lib/crawlers/activeAccounts.js]: I think this fix/improvement should go to a different PR. Otherwise, it is too many changes in one and while we discuss the rest we cannot merge this simple thing

File: lib/crawlers/blockHarvester.js [hunk 4]
```diff
@@ -54,8 +53,11 @@ module.exports = {
     //
     console.log(`[PolkaStats backend v3] - Block harvester - \x1b[32mAdded ${addedBlocks} blocks in ${((endTime - startTime) / 1000).toFixed(0)}s\x1b[0m`);
     console.log(`[PolkaStats backend v3] - Block harvester - \x1b[32mNext execution in 60m...\x1b[0m`);
-    setTimeout(module.exports.blockHarvester(api, pool, config), POLLING_TIME);
-    
+
+    setTimeout(
+      () => module.exports.blockHarvester(api, pool, config),
+      config.pollingTime,
+    );
```
Comment (Bitcoinera) [inline lib/crawlers/blockHarvester.js]: Same than above

File: lib/BackendV3.js (hunk 2 above)
Comment (Bigomby) [reply lib/BackendV3.js]: We could store all the crawlers in an array and loop over it

File: backend.config.js [hunk 5]
```diff
@@ -1,21 +1,34 @@
+// Also wss://kusama-rpc.polkadot.io
+const DEFAULT_WS_PROVIDER_URL = "ws://kusama-rpc.polkadot.io";
```
Comment (Bitcoinera) [inline backend.config.js]: Yeeeey!!! :tada: Finally

File: backend.config.js [hunk 6]
```diff
@@ -1,21 +1,34 @@
...
+  wsProviderUrl: process.env.WS_PROVIDER_URL || DEFAULT_WS_PROVIDER_URL,
+
   postgresConnParams: {
-    user: 'polkastats',
-    host: 'postgres',
-    database: 'polkastats',
-    password: 'polkastats',
-    port: 5432,
+    user: "polkastats",
+    host: "localhost",
```
Comment (Bitcoinera) [inline backend.config.js]: Bring it on bring it on!!! :fire: :fire: :fire:

File: backend.config.js (hunk 1 above)
Comment (Bigomby) [reply backend.config.js]: Done

File: lib/crawlers/activeAccounts.js (hunk 3 above)
Comment (Bigomby) [reply lib/crawlers/activeAccounts.js]: Done

**Comment by Bigomby** [comment]
> Is it wss ??

Fixed

File: backend.config.js [hunk 7]
```diff
@@ -1,21 +1,34 @@
...
+  wsProviderUrl: process.env.WS_PROVIDER_URL || DEFAULT_WS_PROVIDER_URL,
+
   postgresConnParams: {
-    user: 'polkastats',
-    host: 'postgres',
-    database: 'polkastats',
-    password: 'polkastats',
-    port: 5432,
+    user: "polkastats",
+    host: "postgres",
```
Comment (Bitcoinera) [inline backend.config.js]: Nooooo :crying_cat_face:

File: backend.config.js (hunk 5 above)
Comment (Bigomby) [reply backend.config.js]: Sorry, actually it was a mistake. I didn't intend to make this change.

File: lib/BackendV3.js (hunk 2 above)
Comment (Bitcoinera) [reply lib/BackendV3.js]: Let's do that :dancer:

File: backend.config.js (hunk 7 above)
Comment (Bigomby) [reply backend.config.js]: Sorry about creating expectations

File: backend.config.js (hunk 5 above)
Comment (Bitcoinera) [reply backend.config.js]: Small thing. There is an `s` missing in `wss` here

File: lib/BackendV3.js (hunk 2 above)
Comment (mariopino) [reply lib/BackendV3.js]: Looks great! Good idea, I think we can use and array for the crawlers and do a foreach to start them based on conf.

File: backend.config.js (hunk 7 above)
Comment (Bitcoinera) [reply backend.config.js]: It's fine. In the end victory will be ours! :muscle:

File: lib/crawlers/activeAccounts.js (hunk 3 above)
Comment (mariopino) [reply lib/crawlers/activeAccounts.js]: Needed for the pr scope I think

Comment (Bitcoinera) [reply lib/crawlers/activeAccounts.js]: The `config.pollingTime` yes indeed. It is fine @Bigomby . You can leave this here

Comment (Bigomby) [reply lib/crawlers/activeAccounts.js]: Too late :rofl:

Comment (mariopino) [reply lib/crawlers/activeAccounts.js]: nooooo, please restore :-)

Comment (mariopino) [reply lib/crawlers/activeAccounts.js]: And setTimeout sintax is cleaner!

File: backend.config.js (hunk 1 above)
Comment (mariopino) [reply backend.config.js]: dont see yet!

Comment (Bigomby) [reply backend.config.js]: https://github.com/Colm3na/polkastats-backend-v3/pull/69/files#diff-727a97cd278f2750a0af301c4beb6f6a

File: lib/crawlers/activeAccounts.js (hunk 3 above)
Comment (Bigomby) [reply lib/crawlers/activeAccounts.js]: Restored

File: backend.config.js (hunk 1 above)
Comment (mariopino) [reply backend.config.js]: But we can store all the crawlers in an array and loop over it as you suggested ^^

Comment (Bigomby) [reply backend.config.js]: Done

File: backend.config.js [hunk 8]
```diff
@@ -1,21 +1,58 @@
...
+    {
+      enabled: true,
+      module: require('./lib/crawlers/activeAccounts.js'),
+      config: {
+        pollingTime: 1 * 60 * 1000,
+      },
+    },
+
+    {
+      enabled: false,
```
Comment (Bitcoinera) [inline backend.config.js]: 😿
